tpl/template_funcs.go

```diff
@@ -1392,6 +1392,44 @@ func replace(a, b, c interface{}) (string, error) {
 	return strings.Replace(aStr, bStr, cStr, -1), nil
 }
 
+// partialCache represents a cache of partials protected by a mutex.
+type partialCache struct {
+	mu sync.RWMutex
+	p  map[string]template.HTML
+}
+
+// Get retrieves partial output from the cache based upon the partial name.
+// If the partial is not found in the cache, the partial is rendered and added
+// to the cache.
+func (c *partialCache) Get(key, name string, context interface{}) (p template.HTML) {
+	var ok bool
+
+	c.mu.Lock()
```

Comment:
By double-checking i meant ... double. In this case you could use a regular mutex -- but I'm pretty sure proper use of a rwmutex will get much faster in this case.

So:

```
RLock
p, ok = m[key]
RUnlock

if ok return p

Lock

// double check
if p, ok = m[key]

else p = partial ...
Unlock

```